### PR TITLE
Add GraalJS support for conditional authentication functions

### PR DESCRIPTION
**Pull request by shanggeeth**
### Purpose
* Introduce GraalJS proxy objects.
* Add wrapper factory for GraalJS proxy objects.
* Modify adaptive funtions to support GraalJs Script engine execution
* Add unit tests for GraalJs based script engine

### Depends on:
- https://github.com/wso2/carbon-identity-framework/pull/5279

### CODE REVIEW

File: components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/test/resources/testng.xml
Comment (shanggeeth) [inline components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/test/resources/testng.xml:2]: Check the new licences

File: components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/test/resources/testng.xml
Comment (shanggeeth) [inline components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/test/resources/testng.xml:28]: Conditionally run the graaljs related tests

File: components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/test/java/org/wso2/carbon/identity/conditional/auth/functions/analytics/CallAnalyticsFunctionImplTest.java
Comment (shanggeeth) [inline components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/test/java/org/wso2/carbon/identity/conditional/auth/functions/analytics/CallAnalyticsFunctionImplTest.java:77]: Remove the comment

File: components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/test/resources/testng.xml
Comment (shanggeeth) [inline components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/test/resources/testng.xml:34]: Add new line

File: components/org.wso2.carbon.identity.conditional.auth.functions.choreo/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/choreo/CallChoreoFunctionImpl.java
Comment (shanggeeth) [inline components/org.wso2.carbon.identity.conditional.auth.functions.choreo/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/choreo/CallChoreoFunctionImpl.java:131]: Add a comment explaining why we're cloning the parameters

File: components/org.wso2.carbon.identity.conditional.auth.functions.elk/src/test/resources/testng.xml
Comment (shanggeeth) [inline components/org.wso2.carbon.identity.conditional.auth.functions.elk/src/test/resources/testng.xml:34]: Add a new line

File: components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/SetCookieFunctionImpl.java
Comment (shanggeeth) [inline components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/SetCookieFunctionImpl.java:40]: Add class comments

File: components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/HTTPPostFunctionImpl.java
Comment (shanggeeth) [inline components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/HTTPPostFunctionImpl.java:30]: Check what's the minimal set of JARs required.

File: components/org.wso2.carbon.identity.conditional.auth.functions.user/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/user/model/graaljs/JsGraalApplication.java
Comment (shanggeeth) [inline components/org.wso2.carbon.identity.conditional.auth.functions.user/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/user/model/graaljs/JsGraalApplication.java:2]: Change the license

File: components/org.wso2.carbon.identity.conditional.auth.functions.user/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/user/model/graaljs/JsGraalUserSession.java
Comment (shanggeeth) [inline components/org.wso2.carbon.identity.conditional.auth.functions.user/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/user/model/graaljs/JsGraalUserSession.java:65]: Don't need to add this.

File: components/org.wso2.carbon.identity.conditional.auth.functions.user/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/user/model/graaljs/JsGraalUserSession.java
Comment (shanggeeth) [inline components/org.wso2.carbon.identity.conditional.auth.functions.user/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/user/model/graaljs/JsGraalUserSession.java:90]: Remove the new line

File: pom.xml
Comment (shanggeeth) [inline pom.xml:492]: Update the version once the framework PR is merged.

File: components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/HTTPPostFunctionImpl.java
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/HTTPPostFunctionImpl.java:30]: This only requires the graal-sdk to be packed which is ~700 KB in Size.

File: components/org.wso2.carbon.identity.conditional.auth.functions.user/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/user/model/graaljs/JsGraalUserSession.java
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.conditional.auth.functions.user/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/user/model/graaljs/JsGraalUserSession.java:65]: Need to add this since the method needs to be implemented.

File: components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/test/resources/testng.xml
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/test/resources/testng.xml:28]: We can run these tests regardless of graaljs being a connector. Hence keeping these tests.

File: components/org.wso2.carbon.identity.conditional.auth.functions.user.store/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/user/store/UserStoreFunctions.java
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.conditional.auth.functions.user.store/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/user/store/UserStoreFunctions.java]: This will not break the existing logic as existing logic also requires Context to be there. 
However a new functional interface is added for the new method signature.

File: pom.xml
Comment (inthirakumaaran) [reply pom.xml]: Will approve after this

**Comment by shanggeeth** [comment]
Unit test case failures since the PR depends on
- https://github.com/wso2/carbon-identity-framework/pull/5279

Product build is locally successful with
- https://github.com/wso2/carbon-identity-framework/pull/5279
- https://github.com/wso2/product-is/pull/19434

Screenshot 2024-05-20 at 17 43 59

File: components/org.wso2.carbon.identity.conditional.auth.functions.user/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/user/model/graaljs/JsGraalUserSession.java
Comment (inthirakumaaran) [inline components/org.wso2.carbon.identity.conditional.auth.functions.user/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/user/model/graaljs/JsGraalUserSession.java:2]:
```suggestion
 * Copyright (c) 2024, WSO2 LLC. (http://www.wso2.com).
```

**Review by inthirakumaaran — DISMISSED** [review]

**Review by ZiyamSanthosh — APPROVED** [review]

**Pull request merged**
